cmd/neofs-node/policy.go

```diff
@@ -2,26 +2,113 @@ package main
 
 import (
 	"fmt"
+	"sync"
 
+	"github.com/hashicorp/golang-lru/v2/simplelru"
 	"github.com/nspcc-dev/neofs-node/pkg/core/container"
 	"github.com/nspcc-dev/neofs-node/pkg/core/netmap"
 	cid "github.com/nspcc-dev/neofs-sdk-go/container/id"
 	netmapsdk "github.com/nspcc-dev/neofs-sdk-go/netmap"
 )
 
+// storagePolicyRes structures persistent storage policy application result for
+// particular container and network map incl. error.
+type storagePolicyRes struct {
+	nodeSets [][]netmapsdk.NodeInfo
+	err      error
+}
+
+// containerNodesAtEpoch is a thread-safe LRU cache mapping containers into
+// storage policy application results for particular network map.
+type containerNodesAtEpoch struct {
+	mtx sync.RWMutex
+	lru simplelru.LRUCache[cid.ID, storagePolicyRes]
+}
+
+// containerNodesCaches groups caches for all tracked epochs.
+type containerNodesCaches struct{ cur, prev *containerNodesAtEpoch }
```

Comment:
We can avoid having cur/prev here with `[cid, epoch]` key.

Reply:
does not worth it for exactly 2 epochs imo. Moreover, we are likely to decide having less capacity for previous epoch as less in demand. This approach allows such flexibility, while the total number of items will be the same

Reply:
But with `[cid, epoch]` key you may not worry about magic 2-3-5-10 epochs at all, it'll all be naturally added/evicted as time goes.

Reply:
>  naturally added/evicted as time goes

i dont think this is good. We know exactly when some results become stale, and we should free them for GC at this point

Reply:
It's not worth the additional management code. It's a very limited amount of data and combined key simplifies this cache a lot.

Reply:
> It's not worth the additional management code

on the contrary, the code does not change much while respecting the expected behavior. App proceeds only two latest epochs, caching older results and hope they will be evicted by N new requests is incorrect to me. Previously, app didnt cache stale items. And now let it be? Nonsense

Reply:
redone